### PR TITLE
feat: [CDS-51743]: added code for bamboostep

### PR DESCRIPTION
**Pull request by sahithikolichala**
#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by sahithikolichala** [comment]
retrigger coverage

**Comment by sahithikolichala** [comment]
retrigger jest

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    if (this.isTemplatizedView(stepViewType)) {
+      return (
+        <BambooStepInputSet
+          initialValues={initialValues}
+          template={inputSetData?.template}
+          path={inputSetData?.path || ''}
+          inputSetData={inputSetData}
+          readonly={!!inputSetData?.readonly}
+          stepViewType={stepViewType}
+          onUpdate={(values: any) => onUpdate?.(this.processFormData(values))}
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx:192]:
```suggestion
          onUpdate={(values: BambooStepData) => onUpdate?.(this.processFormData(values))}
```

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+          onChange={(values: any) => onChange?.(this.processFormData(values))}
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx:193]:
```suggestion
          onChange={(values: BambooStepData) => onChange?.(this.processFormData(values))}
```

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+          onUpdate={(values: any) => onUpdate?.(this.processFormData(values))}
+        />
+      )
+    }
+
+    return (
+      <BambooStepBaseWithRef
+        initialValues={initialValues}
+        allowableTypes={allowableTypes}
+        onChange={(values: any) => onChange?.(this.processFormData(values))}
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx:212]:
```suggestion
        onChange={(values: BambooStepData) => onChange?.(this.processFormData(values))}
```

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        stepViewType={stepViewType || StepViewType.Edit}
+        onUpdate={(values: any) => onUpdate?.(this.processFormData(values))}
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStep.tsx:214]:
```suggestion
        onUpdate={(values: BambooStepData) => onUpdate?.(this.processFormData(values))}
```

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepBase.tsx [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+              ) {
+                return
+              }
+              refetchPlans()
+            },
+            allowableTypes
+          }}
+        />
+        {getMultiTypeFromValue(formik.values.spec.planName) === MultiTypeInputType.RUNTIME && (
+          <ConfigureOptions
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepBase.tsx:244]:
```suggestion
          < SelectConfigureOptions
```
To support allowed values. Check [https://harness.atlassian.net/wiki/spaces/PD/pages/21173404394/Allowed+Values+-+Pipeline+Team+-+Dev+changes](https://harness.atlassian.net/wiki/spaces/PD/pages/21173404394/Allowed+Values+-+Pipeline+Team+-+Dev+changes)

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepBase.tsx [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+                  >
+                    {getString('pipeline.bambooStep.addPlanParameters')}
+                  </Button>
+                </div>
+              )
+            }}
+          />
+        </MultiTypeFieldSelector>
+        {getMultiTypeFromValue(formik.values?.spec?.planParameter as string) === MultiTypeInputType.RUNTIME && (
+          <ConfigureOptions
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepBase.tsx:330]: Check for allowed values support for this field.

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepInputSet.tsx [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+import css from '@pipeline/components/PipelineSteps/Steps/Steps.module.scss'
+
+import stepCss from './BambooStep.module.scss'
+
+export const jobParameterInputType: SelectOption[] = [
+  { label: 'String', value: 'String' },
+  { label: 'Number', value: 'Number' }
+]
+
+function BambooStepInputSet(formContentProps: any): JSX.Element {
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/BambooStepInputSet.tsx:50]: Add proper type here

File: src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/types.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  spec: BambooStepSpec
+}
+
+export interface BambooStepDeploymentModeProps {
+  stepViewType: StepViewType
+  initialValues: BambooStepData
+  allowableTypes: AllowedTypes
+  onUpdate?: (data: BambooStepData) => void
+  inputSetData?: InputSetData<BambooStepData>
+  formik?: any
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/BambooStep/types.tsx:40]: Add proper type here

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx [hunk 9]
```diff
@@ -389,7 +395,11 @@ const Content = (props: BambooRenderContent): React.ReactElement => {
                     : planDetails,
                   usePortal: true,
                   addClearBtn: !readonly,
-                  noResults: <Text lineClamp={1}>{plansError}</Text>,
+                  noResults: (
+                    <Text lineClamp={1} width={500} height={35} padding="small">
+                      {plansError ? get(plansError, 'data.message', '') : 'Cannot Fetch Plans'}
```
Comment (pankajsingh-harness) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx:400]: Move this hard coding to yaml.

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx [hunk 10]
```diff
@@ -453,7 +463,7 @@ const Content = (props: BambooRenderContent): React.ReactElement => {
                   ]),
                   noResults: (
                     <Text lineClamp={1} width={500} height={35} padding="small">
-                      {artifactPathError}
+                      {artifactPathError ? get(artifactPathError, 'data.message', '') : 'Cannot Fetch Artifact Paths'}
```
Comment (pankajsingh-harness) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx:466]: Same as above

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx [hunk 11]
```diff
@@ -506,7 +516,7 @@ const Content = (props: BambooRenderContent): React.ReactElement => {
                   ]),
                   noResults: (
                     <Text lineClamp={1} width={500} height={35} padding="small">
-                      {buildError}
+                      {buildError ? get(buildError, 'data.message', '') : 'Cannot Fetch Builds'}
```
Comment (pankajsingh-harness) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/BambooArtifactSource/BambooArtifactSource.tsx:519]: Same as above

**Review by pankajsingh-harness — CHANGES_REQUESTED** [review]

File: src/modules/70-pipeline/strings/strings.en.yaml [hunk 12]
```diff
@@ -1713,3 +1714,4 @@ bambooStep:
   addPlanParameters: Add Plan Parameter
   noPlans: No Plans Defined
 namespaceValidation: Namespace must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc')
+cannotFetchOptions: Loading {{fieldName}}...
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/strings/strings.en.yaml:1717]: We have this `loadingFieldOptions: Loading {{fieldName}}s...` in `src/modules/10-common/strings/strings.en.yaml`

**Review by pankajsingh-harness — APPROVED** [review]

**Pull request merged**